src/c-writer.cc

```diff
@@ -4819,23 +4948,45 @@ void CWriter::WriteCHeader() {
 }
 
 void CWriter::WriteCSource() {
-  stream_ = c_stream_;
+  /* Write the "top" to h_impl stream */
+  stream_ = h_impl_stream_;
   Write("/* Automatically generated by wasm2c */", Newline());
   WriteSourceTop();
-  WriteFuncTypes();
+
+  /* Write module-wide declarations to impl header */
+  WriteFuncTypeDecls();
   WriteTagTypes();
-  WriteTags();
+  WriteTagDecls();
   WriteFuncDeclarations();
+  WriteDataInitializerDecls();
+  WriteElemInitializerDecls();
+
+  /* Write #include impl header to every output stream if there are multiple .c
+   * outputs*/
+  if (c_streams_.size() > 1) {
+    assert(header_impl_name_.size() > 0);
+    for (auto& stream : c_streams_) {
+      stream_ = stream;
+      Write("#include \"", header_impl_name_, "\"", Newline());
```

Comment:
This may be kind of annoying, but I ran some quick tests with `run-spec-wasm2c.py -j64`, and the overhead of running GCC with optimization on a ton of C outputs where the only contents are `#include "[test].0-impl.h"` is fairly significant. :-( (E.g. the try-catch test takes about 11.5 seconds on my laptop with `-j64`, compared with 1.7 seconds if I manually delete the `#include` line from the empty files. Those are both running GCC serially 64x3=192 times, most of which are on near-empty source files. Versus... about 0.5 seconds in single-output mode.)

Is there any way we can gracefully defer printing the `#include [modname].0-impl.h` line until the first function is being printed to that output stream? (Maybe we can check the stream offset in `WriteFuncs`?) Might need some care if we're also trying to write the `/* Automatically generated */` header. :-/

Reply:
Then, this would be like if some function is printed to a certain .c output, we need to write `/*Automatically generated*/` header and `#include -impl.h` to the .c output.

Reply:
Yeah, that makes sense to me.

Reply:
It occurs to me that I guess formally it's invalid to have a completely empty compilation unit in C (https://stackoverflow.com/questions/26541150/warning-iso-c-forbids-an-empty-translation-unit). This is triggering a warning/error in GCC when I try to compile with `-Wpedantic`.

What do you think is the best way to handle this? I guess we could add a pass at the end over all the streams and, I guess, if they're completely empty, maybe print something like `/* Empty wasm2c generated file */` and then a dummy typedef?